CRM/Extension/Browser.php

```diff
@@ -157,21 +157,31 @@ public function getExtensions() {
 
   /**
    * Get a description of a particular extension.
+   * This function is an optimised form of getExtensions() - it returns the
+   * extension's details as soon as it's found without caching anything.
    *
    * @param string $key
    *   Fully-qualified extension name.
    *
    * @return CRM_Extension_Info|NULL
    */
   public function getExtension($key) {
-    // TODO optimize performance -- we don't need to fetch/cache the entire repo
-    $exts = $this->getExtensions();
-    if (array_key_exists($key, $exts)) {
-      return $exts[$key];
+    if (!$this->isEnabled() || count($this->checkRequirements())) {
+      return NULL;
     }
-    else {
+
+    $remote = $this->_discoverRemote();
```

Comment:
Sounds fair to say that `discoverRemote()` could be optimized, either by removing the property `_remotesDiscovered` or by adding a timestamp on it so it could also be checked if it is stale or not and only fetch a fresh set if it is. But, as you mentioned it might not be used more than once, so it seems better to just get rid of the property.

Regarding the optimization of `getExtension()` in this PR: I still think the small performance benefit that you get from returning when the loop finds the matching extension (instead of creating an array from all of them first) is not really worth the extra code, and something like: 

```
public function getExtension($key) {
    return CRM_Utils_Array::value($key, $this->getExtensions());
}
```

(untested code) would be cleaner.